### PR TITLE
Fixed gradients of mean and sum

### LINKED ISSUE #74: Gradient wrt `mean` incorrect when axes specified
When taking the mean of a Tensor `t` over a subset of its axes (i.e. producing a non-scalar output), the resulting gradient of `t` will have an incorrect shape and contain too many entries.

Minimum example:
```
const t = sm.full([2, 3], 1).requireGrad()
const m = t.mean([0], true)
const p = m.matmul(sm.full([3, 1], 10))
p.backward()
console.log(p.grad.shape)  // [] correct
console.log(m.grad.shape)  // [1, 3] correct
console.log(t.grad.shape)  // [2, 9] wrong
```

The gradient function seems to be assuming that the incoming gradient will always be a scalar, and simply tiling the incoming gradient using the shape of the forward input Tensor.

Without the `matmul` step above, the incoming gradient will be scalar, and the resulting gradient for `t` would have the correct shape.

### PR DESCRIPTION
**Pull request by yushiyangk**
Previously, these gradient functions assumed that the forward function was applied over all axes, and simply tiled the `grad_in` to the same shape as the forward input Tensor. This breaks when the forward function was applied over only a subset of axes, whereupon `grad_in` would not be a scalar and would have the wrong shape when tiled.

This fix was needed in order for the gradients of Transformer-related modules (#78) to behave correctly. This also fixes #74.

Changed `traverse_gradients` and `async_traverse_gradients` to throw an error when `tensor.backward()` is invoked on a non-scalar Tensor. This matches the behaviour in PyTorch. Fixed the gradient and tests of `sm.concatenate` to correctly reflect this behaviour.

### CODE REVIEW

File: shumai/tensor/register_gradients.ts [hunk 1]
```diff
@@ -10,6 +10,18 @@ export interface Grad {
...
+function recoverShape(tensor: Tensor, originalShape: number[], lostAxes: number[]) {
+  const shapeForBroadcast = [...originalShape]
+  for (let axis of lostAxes) {
+    if (axis < 0) {
+      axis += originalShape.length
+    }
+    shapeForBroadcast[axis] = 1
+  }
+  const tensorForBroadcast = tensor.reshape(shapeForBroadcast)
+  return tensorForBroadcast.add(sm.full(originalShape, 0))
```
Comment (bwasti) [inline shumai/tensor/register_gradients.ts:22]: @jacobkahn @StrongerXi, do y'all know if this will likely be JIT compiled away? (addition of 0 to induce broadcast)

File: shumai/tensor/tensor.ts [hunk 2]
```diff
@@ -214,13 +232,15 @@ export function backward(base_t: Tensor, jacobian: Tensor) {
...
+  const calc_grads = (
+    all_grads_dict: Record<number, [Tensor, Tensor]>
+  ): Record<string, { grad: Tensor; tensor: Tensor }> => {
     const all_grads: Record<string, { grad: Tensor; tensor: Tensor }> = {}
     for (const key in all_grads_dict) {
       const [t, g] = all_grads_dict[key]
       // NB: not really safe in parallel, but convenient for stateful API
       t.grad = g
-      all_grads[t] = {
+      all_grads[t.ptr] = {
```
Comment (bwasti) [inline shumai/tensor/tensor.ts:243]: why's this changed?  I believe `t` will be implicitly converted toString() (which uses the ptr)

File: shumai/tensor/tensor.ts [hunk 3]
```diff
@@ -594,6 +614,10 @@ export class Tensor {
       ...arrayArg(stride, FFIType.i64)
     )
   }
+
+  softmax(axis: number): Tensor {
```
Comment (bwasti) [inline shumai/tensor/tensor.ts:618]: landed the other change - please rebase when you get a chance

File: test/concatenate.test.ts [hunk 4]
```diff
@@ -257,24 +257,4 @@ describe('concatenate', () => {
     expect(isShape(a.grad, expectedGradShape)).toBe(true)
     expect(isShape(b.grad, expectedGradShape)).toBe(true)
   })
-  it('gradient from mean', () => {
```
Comment (bwasti) [inline test/concatenate.test.ts:260]: why delete this?

File: shumai/tensor/register_gradients.ts (hunk 1 above)
Comment (StrongerXi) [reply shumai/tensor/register_gradients.ts:22]: It'll be pretty easy to pick up and rewrite such pattern in the JIT I'm working on.

Could you give an example of this add op (with input shapes)?

If I understand this correctly, the goal is to potentially expand `tensorForBroadcast` in some axes, which sounds like `fl::tile`?

```
Tensor t = rand(Shape(1, 3, 1));

t + full(0, Shape(3, 3, 3))
===>
fl::tile(t, Shape(3, 1, 3))
```

File: test/concatenate.test.ts (hunk 4 above)
Comment (yushiyangk) [reply test/concatenate.test.ts:260]: I actually just moved it into the other test, replacing the other one which used to be just `concatenate` without a `mean`. I realised that that shouldn't be correct, because we should only be calling `backward` on scalars.

File: shumai/tensor/tensor.ts (hunk 2 above)
Comment (yushiyangk) [reply shumai/tensor/tensor.ts:243]: VS Code was not happy about the implicit conversion, so I thought I could make it clearer by making it explicit. It did just occur to me that I've made it a `number` instead of `string`; not sure what I was thinking.

I could restore it back to just `t`, or make it an explicit string conversion instead?

Comment (yushiyangk) [reply shumai/tensor/tensor.ts:243]: Come to think of it, why not index it with its address like in `all_grads_dict`? Although I am not sure what the output object here gets used for.

File: shumai/tensor/register_gradients.ts (hunk 1 above)
Comment (yushiyangk) [reply shumai/tensor/register_gradients.ts:22]: Hi, yes that's what I was trying to do. I was trying to skip the extra logic for determining the tile shape.

Comment (bwasti) [reply shumai/tensor/register_gradients.ts:22]: @jacobkahn - this seems like an FL bug?

@yushiyangk - I'd avoid the `deepCopy()` for now and leave the impl as is :)

File: shumai/tensor/tensor.ts (hunk 2 above)
Comment (bwasti) [reply shumai/tensor/tensor.ts:243]: good question! the idea is to return a map from Tensors to gradients:

```javascript
const grad_map = loss.backward()
const grad_i_care_about = grad_map[weight].grad
```

This is to allow two different backward passes to run at the same time without both overriding `weight.grad` (an issue with the PyTorch approach)

Comment (bwasti) [reply shumai/tensor/tensor.ts:243]: I think your best bet is to make it an explicit conversion to string to make VSCode happy

File: shumai/tensor/register_gradients.ts (hunk 1 above)
Comment (yushiyangk) [reply shumai/tensor/register_gradients.ts:22]: Got it, I've updated the PR with the original version of `recoverShape`, thanks!

**Review by bwasti — APPROVED** [review]
